watchers: use singleton server

marechal-p opened the pull request:
Theia spawns as many nsfw servers as there are clients, maybe more.

This commit replaces this behaviour by centralizing everything into the
same service process in order to optimize resources (watch handles are a
limited resource on a given Linux host).

Some notes on the current design:
- Every time someone requests to watch an `(uri, options)` couple we will
re-use the same watch handle (and actual nsfw instance). This allows for
resource allocation optimizations.
- Since a given service can only serve a single client, I added an event
dispatcher named `FileSystemWatcherServerDispatcher`. You must register
yourself with some unique id you will use to make requests.
- The new `.watchUriChanges` function now takes a `clientId` so that
when events are sent to the dispatcher it then gets routed back to your
own client.
- Added a `--nsfw-watcher-verbose` backend argument to enable verbose
logging for the watcher server.

Signed-off-by: Paul Maréchal <paul.marechal@ericsson.com>

<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### How to test

- Enable the `sample.file-watching.verbose` preference.
- Run `yarn --cwd examples/browser start --nsfw-watcher-verbose`
- Watch for the different logs coming out of the watcher server:
  - There shouldn't be new watchers without a frontend opened.
  - Once you open a frontend on a given workspace for the first time, you should see `NEW` log notifications.
  - If you open a second frontend on the same workspace, you should see `REF++` log notifications
  - Closing one of the two frontend should emit `REF--` log notifications.
  - Closing the last frontend should schedule the watcher de-allocation about 10s later.
    - But if you re-open a frontend before 10s, it should reference the watchers again and not dispose of them.
- Comparing with master:
  - Run the backend.
  - Run `lsof | grep inotify | wc -l` a couple of times before opening a frontend and note the general value.
  - Open frontends on some big workspace (your home maybe?) while noting the new value for `lsof | grep inotify | wc -l` each time.
  - The conclusion should be that on master the number increases with each frontend, but not with these changes.
- Running the `@theia/example-browser` app:
  - Open 2 tabs on different workspaces.
  - Create and delete files in the respective workspaces.
  - Make sure that a log is only reported from the frontend that was watching the modified workspace.
    - `Sample File Watching: X file(s) changed! <workspace>`

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

marechal-p commented:
This is a draft pull-request, I'd like to know if this change cause trouble to any party.

This is an optimization that should help us at Ericsson handle very large workspaces when users might open multiple windows.

The changes should be fully functional, only the API would require some cleaning: `watchUriChanges2` and the like. I want to know if the shape is good before making anything definitive.

marechal-p commented:
@akosyakov I originally wanted to get a unique identifier for a given frontend connecting to the backend in the `FileSystemWatcherClient` class, but I had to create a sequence manually instead: https://github.com/eclipse-theia/theia/blob/mp/watch-server-singleton/packages/filesystem/src/node/filesystem-watcher-client.ts#L39

Is there a way using inversify to generate one when handling an incoming frontend connection, and pass that around? When I tried I could get anything to be injected correctly.

akosyakov commented:
I won't have time for reviews in this and next months please proceed. File watching we should ensure that clients only get events which they ask for. If this PR improves performance and does not break it then it is fine.  For performance  optimisations PR should provide a reproducible method to collect some numbers before and after. A reviewer should run this method too and decide whether code change is worth improvements.

marechal-p commented:
@akosyakov thanks for your feedback, I'll provide a way to benchmark tomorrow (although one can already get a feel just enabling the `--nsfw-watcher-verbose` flag).

marechal-p commented:
Added more things to test.

akosyakov commented:
Please add tests to check that opening 2 tabs with different folders ensures that each tab does not receive events about workspace of another tab.

marechal-p commented:
@akosyakov thanks to this simple request, it made me notice a bug where events weren't correctly forwarded when running the server in a child process. This is now fixed and I added more things to test in the PR description.

marechal-p commented:
This PR is good for review now.

kittaakos commented on packages/core/src/common/promise-util.ts:
```diff
@@ -59,3 +59,33 @@ export async function retry<T>(task: () => Promise<T>, delay: number, retries: n
     throw lastError;
 }
+
+export class AsyncLock {
```
Where do you use this class?

marechal-p replied:
- https://github.com/eclipse-theia/theia/pull/8546/files#diff-8053c51c6deb015c4dcc93ed38f9daa9R97
- https://github.com/eclipse-theia/theia/pull/8546/files#diff-8053c51c6deb015c4dcc93ed38f9daa9R251

kittaakos replied:
I see. Thanks!

From the use-site:
> Ensures that events are processed in the order they are emitted, despite being processed async.

The cool thing is that we do not have to reimplement the weel, you can use a lib. Can `p-queue` help achieve the same?

marechal-p replied:
I guess I don't need all the features from `p-queue`, so I saw this implementation I made like writing a for loop: most lines are comments to help someone see through the mechanism. If you tell me that you want me to use `p-queue` I will change it.

kittaakos replied:
> I guess I don't need all the features from `p-queue`

You're right, probably you do not need all the features, but it's already in the app no matter you use it or not. 1.2k ⭐ on the GH repo and the 1,788,419 weekly downloads from npmjs suggests it's well-maintained lib that we can use as-is instead of writing additional code.

marechal-p replied:
Removed :+1:

kittaakos commented on packages/filesystem/src/node/disk-file-system-provider.ts:
```diff
@@ -804,18 +804,35 @@ export class DiskFileSystemProvider implements Disposable,
+                if (this.disposed) {
+                    return;
+                }
+                if (typeof this.watcherId !== 'undefined') {
```
Use `this.watcherId !== undefined`

kittaakos commented on packages/filesystem/src/node/filesystem-watcher-dispatcher.ts:
```diff
@@ -0,0 +1,82 @@
+        } else {
+            for (const clientId of clientIds) {
+                const client = this.clients.get(clientId);
+                if (typeof client !== 'undefined') {
```
Why do not you use `if (client)`. I think you're confused with the [`ReferenceError`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Not_defined).
It's a good practice to do something like `typeof process !== 'undefined'` to make sure you do not have a reference error at runtime (because your code runs in browser), but you do not need to do it every single time, such as for this code. `const client` defines the variable.

marechal-p replied:
Right, I can directly test for `client !== undefined`, just an habit of using `typeof`. Will change.

kittaakos commented on packages/filesystem/src/common/filesystem-watcher-protocol.ts:
```diff
@@ -19,6 +19,47 @@ import { JsonRpcServer, JsonRpcProxy } from '@theia/core';
+     * @param options optional parameters.
+     * @returns promise to a unique `number` handle for this request.
+     */
+    watchFileChanges2(clientId: number, uri: string, options?: WatchOptions): Promise<number>;
```
I do not think we need the `2` suffix for the method names, do we?

marechal-p replied:
It helped me avoid a bit of confusion when handling both APIs (with `clientId` and without). I will work without the suffix, do you advise removing it?

marechal-p replied:
Just renamed those components to remove those suffixes :+1:

marechal-p commented:
@kittaakos thanks for your comments! I think I addressed everything :)

kittaakos commented:
> I think I addressed everything :)

Very nice. Thanks for the heads-up, @marechal-p. I will give it a try on Windows this week.

marechal-p commented:
I rebased to resolve the conflict. cc @kittaakos

kittaakos commented:
> I rebased to resolve the conflict. cc @kittaakos

Thanks for the heads-up, @marechal-p. I tried it last week: it worked for me although I did not fully understand the verbose file-watcher logs. Please someone else also take a look into the changeset. Thanks!

vince-fugnitto commented on packages/filesystem/src/node/nsfw-watcher/nsfw-filesystem-service.ts:
```diff
@@ -0,0 +1,481 @@
+        if (revived) {
+            this.onRefsRevive();
+        }
+        this.debug('REF++', `clientId=${clientId}, clientRefs=${refs.value}, totalRefs=${totalRefs}. revived=${revived}`);
```
My only nitpick would be to add the `fsPath` to these messages so we can determine for which file the reference was updated. It was sometimes difficult to make the correlation between `handle id` and the actual file when debugging.

Same for `REF--`.

marechal-p replied:
Changed `this.debug` to show `this.fsPath` always 👍

vince-fugnitto reviewed (approved):
The changes work very well for me and it is definitely an overall improvement to the file watchers 👍  

I verified the source code (offline review), as well as tested as many use-cases and scenarios I could think of to validate the functionality.
